graphcore_cloud_tools/paperspace_utils/symlink_datasets_and_caches.py

```diff
@@ -282,26 +289,36 @@ def download_file(
     config = TransferConfig(max_concurrency=max_concurrency)
     target = Path(file.local_file)
     target.parent.mkdir(exist_ok=True, parents=True)
-    exception = None
-    try:
-        if not use_cli:
-            s3client.download_file(bucket_name, file.s3file, str(target), Config=config)
-        else:
-            cmd = (
-                f"aws s3 --endpoint-url {aws_endpoint} --profile {aws_credential} "
-                f"cp s3://{bucket_name}/{file.s3file}"
-                f" {target}"
-            ).split()
-            print(cmd)
-            out = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
-            if out.returncode != 0:
-                raise S3DownloadFailed(f"Command {cmd} failed with return code {out.returncode}")
-    except Exception as error:
-        exception = error
+    exceptions = []
+
+    for attempt in range(max_attempts):
+        try:
+            if not use_cli:
+                s3client.download_file(bucket_name, file.s3file, str(target), Config=config)
+            else:
+                cmd = (
+                    f"aws s3 --endpoint-url {aws_endpoint} --profile {aws_credential} "
+                    f"cp s3://{bucket_name}/{file.s3file}"
+                    f" {target}"
+                ).split()
+                print(cmd)
+                out = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
+                if out.returncode != 0:
+                    raise S3DownloadFailed(f"Command {cmd} failed with return code {out.returncode}")
+            # successful download - clear failed errors from previous attempts and break
+            exceptions = []
+            break
+        except Exception as error:
+            exceptions.append(error)
+            if attempt + 1 < max_attempts:
+                print(f"Failed to download file {file}. Retrying - attempt {attempt+2}/{max_attempts}...")
```

Comment:
Can you warn the user of all the errors? So that even if there is success in the logs we have a trace of the reason for the retries in the log?

```suggestion
            print(f"[WARNING] Failed to download file {file} with error: {type(error.__name__)} {error}...")
            if attempt + 1 < max_attempts:
                print(f"Retrying download of {file} - attempt {attempt+2}/{max_attempts}...")
```